wordpress-mobile/WordPress-Android
# Perform the WPCOM email+password login in the background

Fixes #6700 

This PR introduces a Service to handle logging in to WPCOM when the email+password flow is used. That's the flow mentioned in the ticket and this PR is on-purpose limited to that flow so the changeset is smaller and more manageable.

There is some code duplication with `LoginBaseFormFragment` (mainly the `onAccountChanged` and `onSiteChanged` listeners) which should be lifted when all login flows switch to using a Service.

This PR builds on top of the Service implementation introduced with #6825. See the FAQ section below for more information.

## F.A.Q

### Why use a service?
A Service is used so the login procedure can continue even if the app is backgrounded. The FluxC messages that are now missed by the screen Fragments are properly received and handled in the Service without interruption.

### AutoForeground class. What black magic is that??
This class monitors whether the Service has bound "clients" or not. If bound clients exist, Service's state updates are delivered via an EventBus message. If no clients are bound, the Service gets promoted to foreground and its updates are forwarded to a status bar notification. As clients get bound or unbound, this class automatically toggles the Service foreground-ness and hides or shows the notifications.

### Why is the Service promoted to a foreground one?
The Service is promoted to a foreground one to make sure the system doesn't kill it, even on Android 8.0.

### Why is a client (LoginEmailPasswordFragment) attaching and detaching in `onResume` and `onPause`?
The client (the app UI essentially) attaches or detaches to denote that it wants to start or stop listening to the Service updates. That's similar to how other Fragments register/de-register from EventBus. The added functionality here is that when detached, updates are forwarded to the status bar notification.

### _Sticky_ EventBus messages? What's the deal there?
The Service is designed to start and then stop itself on success or failure. The app may be in the background when the Service stops and the sticky message makes sure the latest state of the Service is available to pick when the app resumes. The sticky message also plays the role of _state retention_ and the UI doesn't need to save it in `onSaveInstanceState` or similar.

### Any downsides to using a sticky message?
Yes, someone has to cleanup or initialise the state. See usages of the `clearLoginServiceState()` method.

### Any downsides to using a sticky message?
Yes, someone has to cleanup or initialise the state. See usages of the `clearLoginServiceState()` method.

### Why not always `clearLoginServiceState()` on the Fragment's start?
We want the client (Fragment in this case) to pick up the current state of the Service. Clearing it always when starting breaks the pattern.

### Why is the "logging in" notification sticky?
That notification is active when the Service is in the foreground and its work hasn't finished yet. Android API demands the notification to be sticky under those conditions.

### Why are the error or success notifications dismissible?
The Service is finished by that time. Failure or success, nothing is running in the background anymore and the sticky notification doesn't make sense. User can dismiss the notification and resume the app whenever they prefer to continue.

## To test:
* With the app data cleaned, launch the app
* Tap on "LOG IN", input your WPCOM email in the email screen and tap NEXT
* Tap on "Enter your password instead" link to drop to the password input screen
* Put in your password, tap NEXT and *immediately hit "Home" to exit the app*
* A notification should appear. It would show "Logging in..." or a success/failure message depending on how the login process went. For example, try to put in a wrong password in the previous step.
* If you try this on a 2FA enabled account, the notification will inform you of the fact that a 2FA is needed to continue.
* Tap on the notifications or perhaps resume the app from the launcher as normal. In any case the notification should hide.

The following flow is "touched" by this PR and should continue to work as normal:
* Social login to a 2FA enabled WPCOM account that was not connected before (web nor mobile)

/cc @mzorz you might want to have a look at this. It's based on the solution we reviewed last week and builds on top of it. Thanks!

/cc @theck13 there are some Social login things touched here so, I would appreciate a pass to make sure nothing broke there. Thanks!
